### ISSUE TITLE
--publish-all's "random" port publishing... isn't

### ISSUE THREAD

**Issue by mpalmer** [moby/moby#34310]
**Description**

The `docker-run` man page I'm looking at describes the `--publish-all` option as follows:

> Publish all exposed ports to random ports on the host interfaces.
> [...]
> When using -P, Docker will bind any exposed port to a random port on the host within an ephemeral port range defined by `/proc/sys/net/ipv4/ip_local_port_range`.

However, the observed behaviour is, instead, that ports are allocated sequentially from the bottom of `ip_local_port_range`, with the sequence being reset whenever `dockerd` is restarted.

Most definitions of "random" I can find mention "unpredictability", "governed by or depending on chance", or "in no particular order" as being defining characteristics; "start from X and work your way up one by one" doesn't meet any definition of "random" I can find.

**Steps to reproduce the issue:**

1. Restart `dockerd`, just to get a clean slate.
2. Run `docker run --rm -it --expose 4242 -P alpine sh`
3. Run `docker ps` in another shell, and note the host port number assigned to the container.  It will almost certainly be the first value in `/proc/sys/net/ipv4/ip_local_port_range`.
4. Exit out of that container.
5. Repeat steps 2-4 another dozen times or so.  Note each time that the host port number is almost certainly the next value above the previous one (with a small chance that another process on the machine might have temporarily snaffled a port in the middle).  Continue to repeat until the experiment has sufficient statistical power to convince you this isn't just the cruel hand of random chance toying with your sanity.
6. Restart `dockerd` again.
7. Run `docker run --rm -it --expose 4242 -P alpine sh`
8. Run `docker ps` is another shell, note that the host port has probably gone back to the first value in `/proc/sys/net/ipv4/ip_local_port_range` (or is pretty close to it).

**Describe the results you received:**

Included above.

**Describe the results you expected:**

The host ports allocated should not be allocated sequentially when multiple containers with `-P` are run.

**Additional information you deem important (e.g. issue happens only occasionally):**

The chances of getting the first port in `ip_local_port_range` on my system are about 1 in 28,000.  The chances of getting the next sequential port each time over 10 runs, if my arithmetic is correct, is 1 in 15,992,496,204,377,989,289,565,234,114,532,246,053,888,000.  I don't think I'm that lucky.

On the up side, at least dockerd doesn't reuse recently released ports, like it does with container IP addresses, and it doesn't blindly re-use allocated ports after restart.  So the problem isn't as horrible as it *could* be.  But it's still not great.

Ideally, the relevant code would be fixed to allocate ports randomly, in line with the documentation, but if not, the documentation should be updated to accurately describe the port allocation process.

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:17:22 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:16:12 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 392
Server Version: 17.06.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-3-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 62.36GiB
Name: lolzno
ID: XCA5:PLPV:KBPE:SY63:O254:RZ3S:4UQS:C73M:KYHC:ASCI:25HL:GOQP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: lolzno
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true

WARNING: No swap limit support
```

**Comment by thaJeztah** [comment]
Other than it not being "random"; can you describe what problem you're running into? The intent of the documentation is to mention that (generally) you won't know what port will be published on the host, i.e. there's no _guarantee_ you'll get the same port each time

**Comment by mpalmer** [comment]
The problem I'm running into is that if port assignments aren't random, there's a greatly increased chance that a new container will get the same port as a previous container, and much confusion will result, because as much as we might *like* service discovery data to be absolutely consistent at all times, reality dictates that doesn't always happen.

At any rate, if the documentation's intent is to mention that (generally) you won't know what port will be published on the host, why doesn't it say that?  At least then the imaginative person could understand what they're getting themselves in for, rather than being misled, as is currently the case.

**Comment by thaJeztah** [comment]
If you have suggestions for better wording; contributions to the documentations are always appreciated; perhaps you're interested in opening a pull request to improve that section in the documentation

**Comment by mpalmer** [comment]
You don't want a PR to have the code match the docs?

**Comment by thaJeztah** [comment]
I'll defer that question to @stevvooe @sanimej

**Comment by stevvooe** [comment]
@mpalmer I don't think the word "random" here is meant in the strictest sense. The more accurate term is probably "ephemeral", meaning that you don't control the port allocation.

Likely, this uses the operating system to make port assignments using some linear allocation method. The implementation shouldn't really matter and one should not rely on the behavior. Even if this were truly random, this problem suffers from the birthday paradox, so collisions might be more common than one might think, especially with the small integer space of port allocations. And even if it was truly random, there would still be the chance of failures if ordering affects application behavior.

That said, I think it would be good to reset here and see if we can't address the cause of having port allocation affect your application. If we better understand this, we can possibly address the issue directly and if we do find that the port allocation algorithm is nonideal, we can use your use case as impetus for a real fix.

Specifically, it would be good to know the following:

1. What is your service discovery setup?
2. What errors are you receiving as a result of delayed port information?

Any detail here will help get to the bottom of this and inform a proper change.

**Comment by mpalmer** [comment]
No, this is not a problem to which the birthday paradox applies.  It's an easy mistake to make, but the birthday paradox is only applicable in situations where every pair of elements in a set are being compared against each other.  There is no one set of items which is being pairwise-compared in this case.

If "random" is not an accurate term, it shouldn't be used in the documentation.  I may have mentioned this before.

In answer to your questions:

1. We're using a home-grown solution at the moment, but have observed the same suboptimal behaviour with consul+registrator in the past (with IP addresses rather than ports, but the same fundamental problem applies).  In fact, I don't think there is *any* way a service registration and discovery system *could* guarantee that this problem can't happen, because I don't know of any way a Docker container's destruction (or at least the release of the host port allocation) can be "blocked" until some external process (deregistration in the service discovery system, and updated configurations in service for every service discovery consumer) has completed.

    Even if docker itself *could* be blocked, it's *vanishingly* rare that consumption of service discovery data is completely synchronous, so even if the service registration database is updated, dependent configuration files (say, managed by consul-template) could still be out of date, or the service that consumes the rewritten configuration files may not have loaded the new configuration.

2. If a container A, providing service X, using host port N, is shutdown, and then a new container B, providing service Y, is created, and is then also allocated host port N, any dependent service which still believes service X is on port N will connect to port N, try to talk the protocol of service X, and get confused.

    If you're lucky, service Y will be using a completely different protocol (say, HTTP vs PostgreSQL) and consumers will connect to port N, exchange an incomprehensible stream of gibberish, and disconnect again, without having performed any meaningful action.  Not great, but not terrible.

    The far, *far* more terrifying case is where services X and Y use the same protocol, but are separate service instances (worst are data stores such as Redis servers, which are typically run *sans* authentication, or HTTP app servers running different apps, which are again rarely authenticated)... the resulting :fire: :fire: :fire: is spectacular.  Worse, for applications that tend to cache TCP connections and keep them open for a long period of time, the consumers rarely re-check the service discovery data for changes as long as the TCP connection stays up, so anything connected to the wrong server can stay that way for a *long* time.

Now, I will certainly concede that this problem is *less* of an issue than, say, IP address reuse, because dockerd registers service ports sequentially, and the "counter" only restarts when dockerd is restarted (unlike IP addresses, which get reused as soon as they're free... :rage:).  But if `dockerd` is restarted between the destruction of container A and the creation of container B, and container A happened to have the "first" available port... you're done.

For IP addresses, we worked around the problem by using IPv6 and setting the MAC address (which in turn sets the IPv6 address) based on a hash of the container name and physical host (we name containers for the instance of the service they provide, so there's no chance of collision there), and with a suitable hash function and a 40 bit address space, you don't tend to get collisions.  I'd prefer to not have to go down a similar route for port allocations, but if I do, so be it.  At the very least, though, the documentation should be clear that port allocations are not "random" in any meaningful sense of the word; then those people with sufficient imagination to recognise the implications will be aware of the problem and can devise their own solution.

**Comment by mavenugo** [comment]
@mpalmer Yes. the documentation is incorrect in this case. It is not random and I just looked at the  port mapping logic -
 I went as far back as 1.6 version (https://github.com/moby/moby/blob/v1.6.0/daemon/networkdriver/portallocator/portallocator.go#L173) and the logic has been exactly what you have observed. It is a sequential allocation of port from the `/proc/sys/net/ipv4/ip_local_port_range`. Also as you mentioned, it doesn't reuse the deallocated port.

@thaJeztah At the minimum we should fix the documentation representing this fact so that usecases like @mpalmer brought up are aware of this behavior so that they can take appropriate precautions.

**Comment by thaJeztah** [comment]
Documentation was addressed by https://github.com/moby/moby/pull/35102
